### DISCUSSION TITLE
Is it possible to selectively skip pytest_report_teststatus()?

### DISCUSSION THREAD

**Discussion by ZChristiansen** [pytest-dev/pytest#12601] [Q&A]
This method runs after every test to create a report. I am grabbing information from that report to then send to a few other methods that call APIs to save the test results to a server. This part is working fine; however, I don't always want to send the execution details to the server.

When I run the tests locally, perhaps troubleshooting or creating new tests, I don't want the results to be sent to the server. If the tests are running in the CICD pipeline, I do want them to be sent.

Is it possible to pass a command line argument that can be used to determine whether or not to continue reporting?

Working code
pytest_report_teststatus(report):
   if report.when == "call"
      get_data_from_report()
      call_api1_to_log_to_server(with_data)

What I want
pytest_report_teststatus(report):
   if donot_report is False: # how could I pass this as a command line argument, or as any argument to this method?
      if report.when == "call"
         get_data_from_report()
         call_api1_to_log_to_server(with_data)

run the test
   pytest --no-report

where --no-report ends up setting donot_report = true. I know how to add the command line option using pytest_addoption() and a fixture, I can't figure out how to get that value to pass to pytest_report_teststatus()

**Comment by RonnyPfannschmidt** [comment]
the ask is not clear, the hook itself is part of the common flow, it seems like you want a plugin object you only add to pytest when opted in

**Reply by ZChristiansen** [reply]
The ask is: I want to implement an option on the command line to either send or not send information to a system of record (server) after each test is executed. The pytest_report_teststatus() runs after each test and contains a test report with test information and execution result.

I don't think using Fixtures is going to be able to allow me to toggle the reporting, and the hooks aren't capable of that either. Would a plugin be able to receive command line arguments and have access to the TestReport generated by pytest_report_teststatus()?

**Reply by RonnyPfannschmidt** [reply]
As per https://docs.pytest.org/en/7.1.x/reference/reference.html?highlight=hookspec#pytest.hookspec.pytest_report_teststatus

The hook takes the config 

The config provides access to cli oprions 



Alternatively only register a plugin object when the option is set via the configure hook

**Reply by ZChristiansen** [reply]
Thank you, this has lead me to the correct path. With the below setup in the conftest.py file, I can now call pytest from the command line as 

```pytest --report-tests Y```

And the reporting will occur. Otherwise it will not run the additional steps.

```
def pytest_addoption(parser):
    parser.addoption("--report-tests", action="store", default="N", help="Enable reporting of test execution to Azure Devops. Y or N")


def pytest_configure(config):
    config.report_tests = config.getoption("--report-tests")

@pytest.hookimpl(tryfirst=True)
def pytest_report_teststatus(report, config):
    if config.report_tests == "Y":
        <do reporting things>
```

**Reply by RonnyPfannschmidt** [reply]
Please don't set custom attributes on config

**Reply by ZChristiansen** [reply]
Is there a different location I should set them?

**Reply by RonnyPfannschmidt** [reply]
For your Case the recommended approach is to make a plugin object and registering it in pytest- configure

Then there's no need to check in the hook, as the hook will only register when the option is enabled

**Reply by ZChristiansen** [reply]
What I have works. Can you provide more clarification as to why not to do it this way?

**Reply by RonnyPfannschmidt** [reply]
You dont have ownership on the config object 

If a future change in pytest where to use that attribute name and/or forbid changes to config

Your code would break

**Reply by ZChristiansen** [reply]
If I replace this code
`def pytest_configure(config):
     config.report_tests = config.getoption("--report-tests")`

With this code
```
@pytest.hookimpl(tryfirst=True)
def pytest_collection_modifyitems(session):
    config.report_tests = session.config.getoption("--report-tests")
```
    
Would there still be concerns about ownership of the config object?

**Reply by RonnyPfannschmidt** [reply]
Yes, please use the plugin object examples from the documentation as examples

**Reply by ZChristiansen** [reply]
After some additional research and trial/error, this code works without making changes to config. It uses config.getoption to read the option passed to the command line then performs actions based on that.

```
def pytest_addoption(parser):
    report = parser.getgroup('report')
    report.addoption("--report-tests", action="store", default="N", help="Enable reporting. Y or N")


@pytest.hookimpl(tryfirst=True)
def pytest_report_teststatus(report, config):
    if config.getoption('--report-tests'.upper() == "Y":
        <do reporting things>
```
